torch/csrc/autograd/record_function.h

```diff
@@ -26,32 +28,70 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
-
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
-
-  explicit RecordFunction(
-      std::string name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+  // Default constructor is used with before function called afterwards
+  RecordFunction();
+
+  // before function initializes RecordFunction members and calls
+  // start callbacks;
+  // before function can be used only when RecordFunction is
+  // constructed with default constructor RecordFunction::RecordFunction()
+  void before(const char* name, int64_t sequence_nr = -1);
+  void before(std::string name, int64_t sequence_nr = -1);
+  void before(Function* fn, int64_t sequence_nr = -1);
+
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      int64_t current_sequence_nr,
+      const Args&... args) {
+    if (needs_inputs_) {
+      inputs_ = std::vector<c10::IValue>({args...});
+    }
+    before(fn, current_sequence_nr);
+  }
 
-  explicit RecordFunction(
-      const char* name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+  // When the second argument is not integer, treat it and the rest
```

Comment:
why do you need all this template logic? Since you're constructing a vector of IValues anyway - just move it to the invoking code. I.e. instead of 

```
RECORD_FUNCTION(this, inputs.begin(), inputs.end());
```

can't we just write

```
RECORD_FUNCTION(this, vector<IValue>(inputs.begin(), inputs.end()));
```
explicitly?

Also, that'd resolve your disambiguation problem of int vs seqnum (which feel weird, what if the first argument is an int).

In that case you'd just have 2 versions - `before(name, seq_num=-1, inputs={})` and `before(name, inputs={})`

Reply:
Wow, this is _extremely_ magical. What if the function you're recording takes an integer as the first input? It should never get silently treated as a sequence number. I second what Dmytro said, but you can safely use `ArrayRef<IValue>` instead of vectors

Reply:
@apaszke I explicitly pass only Tensors and Scalars as the inputs, the contract was also described in the comment on before function

@dzhulgakov the template logic was mainly for Args... - so that we don't need to pack the free inputs at the callsite
that is to say we some modifications to the RECORD_FUNCTION macro (as mentioned above) we don't need to use templates